Step5 로또 수동 PR 드립니다.

develo-pyo opened the pull request:
안녕하세요!
로또 수동 PR 요청 드립니다!

궁금한게 있어 질문 드립니다.
수익률 계산과 같은 공식에 대한 테스트가 필요한 경우, 실무에서 보통 테스트 코드를 어떻게 작성하는지 궁금합니다.
메소드 내에서 사용한 공식을 그대로 테스트코드에서 사용하여 expected 값을 만들어 actual 과 비교한다면 테스트에 의미가 없는 듯 한데,
실제로 계산기로 계산해본 값을 expected 로 선언하고 계산 메소드에서 리턴한 actual 값과 비교하는게 맞을까요?

그럼 리뷰 부탁드리겠습니다.
주말 잘보내세요!

develo-pyo committed:
refactor: 정적메소드 사용, 유효성검사 생성자 안으로 이동

develo-pyo committed:
refactor: 당첨여부 판별 메소드 rank enum 내로 이동

develo-pyo committed:
refactor: 로또번호 보너스번호 중복 유효성검사 분리

develo-pyo committed:
docs: step5 수동로또 요구사항 분석

develo-pyo committed:
feat: 수동 로또수 입력 구현

develo-pyo committed:
feat: 수동 로또수 입력값 유효성검사 구현

1. 수동 로또수 입력값 유효성검사 구현
2. 수동 로또수 입력값 유효성검사 테스트 구현

develo-pyo committed:
feat: 로또번호 객체 캐싱 구현

1. 로또번호 객체 캐싱 구현
2. 로또번호 객체 비교 테스트

develo-pyo committed:
feat: 수동로또 생성 구현

develo-pyo committed:
feat: 수익률 계산 버그 수정

bigdecimal 사용하도록 개선

develo-pyo committed:
test: 수익률 계산 테스트 추가

develo-pyo committed:
feat: 구입금액 입력 유효성검사 추가 구현

develo-pyo committed:
docs: 요구사항 구현 확인

wooobo commented on docs/lotto/README.md:
```diff
@@ -55,4 +55,13 @@
 1. 보너스볼 번호를 입력받는다.
```
```
지난 주 당첨 번호를 입력해 주세요.
1,2,3,4,5,5 ==> 중복숫자인데 예외처리 안되어있습니다
보너스 볼을 입력해 주세요.
```

예외 조건을 좀 더 고민해보어야 할것 같아요!

wooobo replied:
# asis
```
구입금액을 입력해 주세요.
1000
수동으로 구매할 로또 수를 입력해 주세요.
5
[ERROR] 구매가능 수량 초과

> Task :LottoMain.main() FAILED
2 actionable tasks: 2 executed
Exception in thread "main" java.lang.IllegalArgumentException: [ERROR] 구매가능 수량 초과
	at lotto.util.InputValidator.validateUserWrittenLottoAmount(InputValidator.java:83)
	at lotto.controller.LottoGame.of(LottoGame.java:26)
	at lotto.LottoMain.main(LottoMain.java:12)

FAILURE: Build failed with an exception.
```

# tobe
```
구입금액을 입력해 주세요.
1000
수동으로 구매할 로또 수를 입력해 주세요.
5
[ERROR] 구매가능 수량 초과
Disconnected from the target VM, address: 'localhost:42693', transport: 'socket'
Connected to the target VM, address: '127.0.0.1:41243', transport: 'socket'
프로그램을 종료합니다.

BUILD SUCCESSFUL in 13s
2 actionable tasks: 2 executed
11:44:49 AM: Execution finished ':LottoMain.main()'.
Disconnected from the target VM, address: '127.0.0.1:41243', transport: 'socket'
```


예외 처리를 하는건 좋지만, 프로그램이 `FAILED`로 강제 종료되는것 보다는 안정적으로 종료되는게 더 좋지 않을까 생각이되어요 :thinking:

develo-pyo replied:
네 수정하겠습니다!

wooobo commented on src/main/java/lotto/constants/Rank.java:
```diff
@@ -33,4 +33,11 @@ public static Rank valueOf(int countOfMatch, boolean matchBonus) {
     }
+
+    public boolean isWinner() {
+        if (this.countOfMatch >= FIFTH.countOfMatch) {
+            return true;
+        }
+        return false;
```
간단하게 인라인 표현 해볼수 있겠네요!

```suggestion
        return this.countOfMatch >= FIFTH.countOfMatch;
```

develo-pyo replied:
넵 수정하겠습니다

wooobo commented on src/test/java/lotto/util/ProfitCalculatorTest.java:
```diff
@@ -13,13 +14,25 @@ class ProfitCalculatorTest {
+        BigDecimal actualProfitRatio = ProfitCalculator.calculateProfitRatio(lottoResult, lottoAmount);
+        BigDecimal expected = new BigDecimal("2000000");
+        assertThat(actualProfitRatio.compareTo(expected) == 0).isTrue();
     }
```
>수익률 계산과 같은 공식에 대한 테스트가 필요한 경우, 실무에서 보통 테스트 코드를 어떻게 작성하는지 궁금합니다.

여쭈어 보시는 이유가, 아무래도 `수익률`을 계산하기 위한 선행해야될 작업이 많은 포인트가 고민이 되는게 맞으실까요?
저는 이렇게 진행했을것 같아요.

### 1. 수익률 기본 코드를 테스트 코드로 구현해보면서, 필요한 데이터 or 로직을 탐색한다.
```
    @Test
    void 수익률_검증() {
        BigDecimal 총구입금액 = BigDecimal.valueOf(10000);
        BigDecimal 총당첨금액 = BigDecimal.valueOf(5000);

        BigDecimal 수익률 = 총구입금액.divide(총당첨금액);

        assertThat(수익률).isEqualTo(BigDecimal.valueOf(2));
    }
```

### 2. 프로덕션 코드에서 해당 데이터를 가지고 있는 객체를 생각해본다.
`LottoResult` => 로또 당첨 총 금액을 알 수 있다.
`Lottos` => 로또 총 구입 갯수를 통해 총 구입금액을 수 할 수 있다.

`LottoResult` 에게 수익률 책임을 위임하는걸로 결정


### 3. `LottoResult` 객체에 코드를 구현해본다.
 `LottoResult`가 총 당첨금액을 알고 있므로, 로또 구입갯수를 전달해본다.
```
    @Test
    void LottoReport_수익률_검증() {
        LottoResult lottoResult = 등수_2등_1개_당첨();

        BigDecimal 수익률 = lottoResult.calculateProfitRatio(5);

        assertThat(수익률).isEqualTo(BigDecimal.valueOf(1));
    }
```

테스트 코드부터 작성한다면, 필요한 코드만 작성하기에 좋을 수 있어요~

wooobo replied:
>메소드 내에서 사용한 공식을 그대로 테스트코드에서 사용하여 expected 값을 만들어 actual 과 비교한다면 테스트에 의미가 없는 듯 한데,
실제로 계산기로 계산해본 값을 expected 로 선언하고 계산 메소드에서 리턴한 actual 값과 비교하는게 맞을까요?

테스트는 `정확한 INPUT` 에 대한 `항상 같은 OUTPUT` 을 검증 하는것이라고도 볼 수 있을것 같습니다. 
계산하는 메서드가 있다면, 그 메서드의 `input`에 대한 `output`을 검증 하는것이므로 `공식`을 `테스트 코드`에서 알 필요는 없습니다. 단지 `해당 메서드`를 사용하는 시나리오(성공케이스,실패케이스 등등)를 테스트 할 뿐이라고 생각해요~

---
ex)
```
int sum(int a, int b) {
  return a+b;
}

void sum_1,2_결과_3() {
  int actual = sum(1,2);
  
  검증(actual).비교(3);
}

void sum_1,2_결과_3() {
  int actual = sum(1,2);
  
  검증(actual).비교(1+2); ==> 테스트 코드를 복잡하게 할 필요가 있을까요? 단지 테스트 에서는 `3`이 나오냐가 중요한게 아닐까요?
}

```

develo-pyo replied:
마지막 예시를 들어주신것처럼 actual 과 expected 를 비교할 때 둘다 같은 수식을 사용하면 이게 테스트가 의미가 있나 싶었습니다.
예상되는 값을 직접 넣어주는게 맞는 방향이군요.  
로또 수익률 계산 메소드도 LottoResult 안쪽으로 옮기면 더 코드가 깔끔해 지겠네요..
자세한 설명과 답변 감사드립니다!! 👍

wooobo reviewed (changes requested):
5단계도 잘진행해주셨습니다 :smile: :+1: 

질문주신 내용과 코멘트 몇가지 작성해두었습니다!

확인 후 재 요청 부탁드려요~ :pray:

develo-pyo committed:
refactor: 당첨 여부 반환 메소드 리팩토링

develo-pyo committed:
fix: 수익률 계산 버그 수정

Non-terminating decimal expansion; no exact representable decimal result. 버그 수정

develo-pyo committed:
fix: 로또번호 범위 초과 유효성 검사 버그 수정

develo-pyo committed:
refactor: input 유효성검사, domain 유효성 검사 분리

develo-pyo committed:
feat: 로또갯수 도메인 구현,

develo-pyo committed:
feat: 로또리스트 생성기 객체 구현

develo-pyo committed:
refactor: 로또 컨트롤러 리팩토링

develo-pyo committed:
refactor: 수익률 계산 메소드 LottoResult 로 이동

wooobo reviewed (approved):
5단계 로또 마지막 단계까지 달려오느라 수고많으셨습니다!!

개인적으로 로또 미션에서 많은 부분을 배울수 있다고 생각하는데요~
시간되시면 다시 한번더 처음부터 해보시는걸 추천해요~

이만 로또 미션 `merge` 하겠습니다!
남은 미션 화이팅입니다!
